core-bundle/src/Routing/Matcher/LegacyMatcher.php

```diff
@@ -0,0 +1,196 @@
+        $this->configAdapter = $this->framework->getAdapter(Config::class);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function matchRequest(Request $request)
+    {
+        $this->framework->initialize();
+
+        if (!isset($GLOBALS['TL_HOOKS']['getPageIdFromUrl']) || !\is_array($GLOBALS['TL_HOOKS']['getPageIdFromUrl'])) {
+            return $this->requestMatcher->matchRequest($request);
+        }
+
+        @trigger_error('Using the "getPageIdFromUrl" hook has been deprecated and will no longer work in Contao 5.0.', E_USER_DEPRECATED);
+
+        $fragments = null;
+
+        if ($this->configAdapter->get('folderUrl')) {
+            try {
+                $match = $this->requestMatcher->matchRequest($request);
+                $fragments = $this->createFragmentsFromMatch($match);
+                $locale = $match['_locale'] ?? '';
+            } catch (ResourceNotFoundException $e) {
+                // Continue and parse fragments from path
+            }
+        }
+
+        if (null === $fragments) {
+            $pathInfo = $this->parseSuffixAndLanguage($request->getPathInfo(), $locale);
+            $fragments = $this->createFragmentsFromPath($pathInfo);
+        }
+
+        if ($this->configAdapter->get('addLanguageToUrl')) {
+            $this->inputAdapter->setGet('language', $locale);
+        }
+
+        $fragments = $this->executeLegacyHook($fragments);
+        $pathInfo = $this->createPathFromFragments($fragments, $locale);
+
+        return $this->requestMatcher->matchRequest(Request::create($pathInfo));
+    }
+
+    private function createFragmentsFromMatch(array $match): array
+    {
+        $page = $match['pageModel'] ?? null;
+        $parameters = $match['parameters'] ?? '';
+
+        if (!$page instanceof PageModel) {
+            throw new ResourceNotFoundException();
+        }
+
+        if ('' === $parameters) {
+            return [$page->alias];
+        }
+
+        $fragments = array_merge([$page->alias], explode('/', substr($parameters, 1)));
+
+        // Add the second fragment as auto_item if the number of fragments is even
+        if ($this->configAdapter->get('useAutoItem') && 0 === \count($fragments) % 2) {
+            array_splice($fragments, 1, 0, ['auto_item']);
+        }
+
+        return $fragments;
+    }
+
+    private function createFragmentsFromPath(string $pathInfo)
+    {
+        $fragments = explode('/', $pathInfo);
+
+        // Add the second fragment as auto_item if the number of fragments is even
+        if ($this->configAdapter->get('useAutoItem') && 0 === \count($fragments) % 2) {
+            array_splice($fragments, 1, 0, ['auto_item']);
+        }
+
+        return $fragments;
+    }
+
+    private function executeLegacyHook(array $fragments)
+    {
+        foreach ($GLOBALS['TL_HOOKS']['getPageIdFromUrl'] as $callback) {
+            /** @noinspection StaticInvocationViaThisInspection */
```

Comment:
I think we have agreed on not adding any IDE specific code. 🙈

Reply:
Removed in 847914e2ed2909575f53e21e66a8e9f42ed40004